### PR TITLE
Added synchronized access to listeners and global subscriptions

### PR DESCRIPTION
**Pull request by jguz-pubnub**
fix(listeners): added synchronized access to listeners and global subscriptions

Added synchronized access inside `WeakSet` and for global subscriptions.

Closes #186

### CODE REVIEW

File: Sources/PubNub/Helpers/WeakBox.swift [hunk 1]
```diff
@@ -29,43 +29,51 @@ final class WeakBox<Element>: Hashable where Element: AnyObject, Element: Hashab
...
+      elements.forEach { element in
+        currentValue.update(with: WeakBox(element))
+      }
+    }
   }
 
   // NSSet Operations
   var allObjects: [Element] {
-    return elements.compactMap { $0.underlying }
+    return elements.lockedRead { $0 }.compactMap { $0.underlying }
```
Comment (parfeon) [inline Sources/PubNub/Helpers/WeakBox.swift:44]: `compactMap` after closure, close will operate on the reference to the original `elements` (they did not mutate in `lockedRead` closure and COW not applied) but outside of lock. The following change should be more thread-safe:
```suggestion
    return elements.lockedRead { $0.compactMap { $0.underlying } }
```

File: Sources/PubNub/Helpers/WeakBox.swift [hunk 2]
```diff
@@ -29,43 +29,51 @@ final class WeakBox<Element>: Hashable where Element: AnyObject, Element: Hashab
...
   }
 
   var count: Int {
-    return self.elements.count
+    elements.lockedRead { $0 }.count
```
Comment (parfeon) [inline Sources/PubNub/Helpers/WeakBox.swift:48]: `count` after closure, close will operate on the reference to the original `elements` (they did not mutate in `lockedRead` closure and COW not applied) but outside of lock. The following change should be more thread-safe:
```suggestion
    elements.lockedRead { $0.count }
```

File: Sources/PubNub/Helpers/WeakBox.swift [hunk 3]
```diff
@@ -29,43 +29,51 @@ final class WeakBox<Element>: Hashable where Element: AnyObject, Element: Hashab
...
-    elements.removeAll()
+    elements.lockedWrite { $0 = Set<WeakBox<Element>>() }
   }
 }
 
 extension WeakSet: Collection {
-  var startIndex: Set<WeakBox<Element>>.Index { return elements.startIndex }
-  var endIndex: Set<WeakBox<Element>>.Index { return elements.endIndex }
+  var startIndex: Set<WeakBox<Element>>.Index { return elements.lockedRead { $0 }.startIndex }
+  var endIndex: Set<WeakBox<Element>>.Index { return elements.lockedRead { $0 }.endIndex }
```
Comment (parfeon) [inline Sources/PubNub/Helpers/WeakBox.swift:70]: `startIndex ` and `endIndex` after closure, close will operate on the reference to the original `elements` (they did not mutate in `lockedRead` closure and COW not applied) but outside of lock. The following change should be more thread-safe:
```suggestion
  var startIndex: Set<WeakBox<Element>>.Index { return elements.lockedRead { $0.startIndex } }
  var endIndex: Set<WeakBox<Element>>.Index { return elements.lockedRead { $0.endIndex } }
```

File: Sources/PubNub/Helpers/WeakBox.swift [hunk 4]
```diff
@@ -29,43 +29,51 @@ final class WeakBox<Element>: Hashable where Element: AnyObject, Element: Hashab
...
 
   subscript(position: Set<WeakBox<Element>>.Index) -> Element? {
-    return elements[position].underlying
+    elements.lockedRead { $0[position] }.underlying
```
Comment (parfeon) [inline Sources/PubNub/Helpers/WeakBox.swift:73]: `underlying` after closure, close will operate on the reference to the original `elements` (they did not mutate in `lockedRead` closure and COW not applied) but outside of lock. The following change should be more thread-safe:
```suggestion
    elements.lockedRead { $0[position].underlying }
```

File: Sources/PubNub/Helpers/WeakBox.swift [hunk 5]
```diff
@@ -29,43 +29,51 @@ final class WeakBox<Element>: Hashable where Element: AnyObject, Element: Hashab
...
   }
 
   func index(after index: Set<WeakBox<Element>>.Index) -> Set<WeakBox<Element>>.Index {
-    return elements.index(after: index)
+    elements.lockedRead { $0 }.index(after: index)
```
Comment (parfeon) [inline Sources/PubNub/Helpers/WeakBox.swift:77]: `index (after:)` after closure, close will operate on the reference to the original `elements` (they did not mutate in `lockedRead` closure and COW not applied) but outside of lock. The following change should be more thread-safe:
```suggestion
    elements.lockedRead { $0.index(after: index) }
```

File: Sources/PubNub/Subscription/SubscriptionSession.swift [hunk 6]
```diff
@@ -163,15 +176,25 @@ class SubscriptionSession: EventEmitter, StatusEmitter {
...
     }
     internalUnsubscribe(
-      from: globalChannelSubscriptions.compactMap { channelNamesToUnsubscribe.contains($0.key) ? $0.value : nil },
-      and: globalGroupSubscriptions.compactMap { groupNamesToUnsubscribe.contains($0.key) ? $0.value : nil },
+      from: globalChannelSubscriptions.lockedRead { $0 }.compactMap {
+        channelNamesToUnsubscribe.contains($0.key) ? $0.value : nil
+      },
+      and: globalGroupSubscriptions.lockedRead { $0 }.compactMap {
+        groupNamesToUnsubscribe.contains($0.key) ? $0.value : nil
+      },
```
Comment (parfeon) [inline Sources/PubNub/Subscription/SubscriptionSession.swift:184]: `compactMap ` after closure, close will operate on the reference to the original `elements` (they did not mutate in `lockedRead` closure and COW not applied) but outside of lock. The following change should be more thread-safe:
```suggestion
      from: globalChannelSubscriptions.lockedRead { $0.compactMap {
        channelNamesToUnsubscribe.contains($0.key) ? $0.value : nil
      } },
      and: globalGroupSubscriptions.lockedRead { $0.compactMap {
        groupNamesToUnsubscribe.contains($0.key) ? $0.value : nil
      } },
```

**Review by parfeon — CHANGES_REQUESTED** [review]

**Comment by jguz-pubnub** [comment]
@pubnub-release-bot release

**Pull request merged**